avalanche/training/templates/online_supervised.py

```diff
@@ -1,209 +1,332 @@
+            pinned memory before returning them. Defaults to True.
+        """
+
+        other_dataloader_args = {}
 
-        for self.n_pass in range(self.num_passes):
-            self.training_epoch(**kwargs)
+        if parse_version(torch.__version__) >= parse_version("1.7.0"):
+            other_dataloader_args["persistent_workers"] = persistent_workers
+
+        self.dataloader = TaskBalancedDataLoader(
+            self.adapted_dataset,
+            oversample_small_groups=True,
+            num_workers=num_workers,
+            batch_size=self.train_mb_size,
+            shuffle=shuffle,
+            pin_memory=pin_memory,
+            **other_dataloader_args
+        )
 
-        # if is_last_sub_exp:
-        self._after_training_epoch(**kwargs)
-        self._after_training_exp(**kwargs)
+    def make_eval_dataloader(
+            self, num_workers=0, pin_memory=True, persistent_workers=False,
+            **kwargs
+    ):
+        """
+        Initializes the eval data loader.
+        :param num_workers: How many subprocesses to use for data loading.
+            0 means that the data will be loaded in the main process.
+            (default: 0).
+        :param pin_memory: If True, the data loader will copy Tensors into CUDA
+            pinned memory before returning them. Defaults to True.
+        :param kwargs:
+        :return:
+        """
+        other_dataloader_args = {}
+
+        if parse_version(torch.__version__) >= parse_version("1.7.0"):
+            other_dataloader_args["persistent_workers"] = persistent_workers
+
+        self.dataloader = DataLoader(
+            self.adapted_dataset,
+            num_workers=num_workers,
+            batch_size=self.eval_mb_size,
+            pin_memory=pin_memory,
+            **other_dataloader_args
+        )
+
+    def forward(self):
+        """Compute the model's output given the current mini-batch."""
+        return avalanche_forward(self.model, self.mb_x, self.mb_task_id)
 
     def model_adaptation(self, model=None):
-        """Adapts the model to the data from the current
-           (full) experience.
+        """Adapts the model to the current data.
 
         Calls the :class:`~avalanche.models.DynamicModule`s adaptation.
         """
         if model is None:
             model = self.model
 
-        for module in model.modules():
-            if isinstance(module, DynamicModule):
-                module.adaptation(self.full_experience)
+        # For training:
+        if isinstance(self.experience, OnlineCLExperience):
+            # If the strategy has access to task boundaries, adapt the model
+            # for the whole origin experience to add the
+            if self.experience.access_task_boundaries:
+                avalanche_model_adaptation(model,
+                                           self.experience.origin_experience)
+            else:
+                self.model_params_before_adaptation = list(model.parameters())
```

Comment:
I don't think this if-then-else is needed. Also notice that it is called at every experience in both cases

Reply:
So there are three possibilities:
During training -> if we have task boundaries, we adapt to the whole original experience at the begggining of the original experience  to avoid model adaptation every sub-experience,
if there we have no task boundaries, we save the model params to call and adapt to `self.experience` to update optimizer later

During evaluation-> since the experience is most probably not an `OnlineCLExperience`, we just adapt the model to `self.experience` without saving current model parameters.

Reply:
What are you trying to achieve exactly? 

There is no difference between doing the adaptation at each step using `self.experience` or doing it on `self.experience.origin_experience`.

Reply:
Ok, I see it now. You need the if because you check the task boundaries _before_ calling `model_adaptation`. IMO you are needlessly complicating the code here because there is no need to distinguish these two cases. Nonetheless, the code seems correct.